src/rust/src/rerr.rs

```diff
@@ -0,0 +1,127 @@
+use extendr_api::{extendr, extendr_module, symbol::class_symbol, Attributes, Rinternals, Robj};
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error)]
+pub enum Rctx {
+    #[error("The argument [{0}] casuse an error because")]
+    BadArgument(String),
+    #[error("Encountered the following error in Extendr:\n{0}")]
+    Extendr(String),
+    #[error("Possibly {0}")]
+    Hint(String),
+    #[error("{0}")]
+    Plain(String),
+    #[error("Encountered the following error in Polars:\n{0}")]
+    Polars(String),
+    #[error("When {0}")]
+    When(String),
+    #[error("Expected a value of type [{0}], but got [{1}] instead")]
+    TypeMismatch(String, String),
+}
+
+#[derive(Clone, Debug)]
+pub struct Rerr(Vec<Rctx>);
+pub type RResult<T> = core::result::Result<T, Rerr>;
+
+pub trait WithRctx<T> {
+    fn ctx(self, rctx: Rctx) -> RResult<T>;
+    fn on_arg<S: Into<String>>(self, arg: S) -> RResult<T>;
+    fn type_mismatch<R: Into<String>, S: Into<String>>(self, ty: R, val: S) -> RResult<T>;
+    fn hint<S: Into<String>>(self, cause: S) -> RResult<T>;
+    fn when<S: Into<String>>(self, env: S) -> RResult<T>;
+}
+
+impl<T, E: Into<Rerr>> WithRctx<T> for core::result::Result<T, E> {
+    fn ctx(self, rctx: Rctx) -> RResult<T> {
+        self.map_err(|e| {
+            let mut rerr = e.into();
+            rerr.0.push(rctx);
+            rerr
+        })
+    }
+
+    fn on_arg<S: Into<String>>(self, arg: S) -> RResult<T> {
+        self.ctx(Rctx::BadArgument(arg.into()))
+    }
+
+    fn type_mismatch<R: Into<String>, S: Into<String>>(self, ty: R, val: S) -> RResult<T> {
+        self.ctx(Rctx::TypeMismatch(ty.into(), val.into()))
+    }
+
+    fn hint<S: Into<String>>(self, cause: S) -> RResult<T> {
+        self.ctx(Rctx::Hint(cause.into()))
+    }
+
+    fn when<S: Into<String>>(self, env: S) -> RResult<T> {
+        self.ctx(Rctx::When(env.into()))
+    }
+}
+
+#[extendr]
+impl Rerr {
+    pub fn new() -> Self {
+        Rerr(Vec::new())
+    }
+
+    pub fn info(&self) -> String {
+        self.0
+            .iter()
+            .rev()
+            .map(|rerr| format!("{}", rerr))
+            .fold(String::from("Error"), |msg, rerr| {
+                format!("{}: {}", msg, rerr)
+            })
+    }
+
+    pub fn contexts(&self) -> Vec<String> {
+        self.0
+            .iter()
+            .rev()
+            .map(|rerr| format!("{:?}", rerr))
+            .collect()
+    }
```

Comment:
it would be needed to add contexts on R side also. I suggest adding `on_arg` like this 

```rust
   pub fn on_arg(&self, s: String) -> Self {
        let res: RResult<i32> = Err(self.clone()).on_arg(s); //i32 is a placeholder type, could be any
        res.unwrap_err()
    }
    ```